internal/controller/nimbuild_controller.go

```diff
@@ -0,0 +1,657 @@
+					newNIMBuild, ok := e.ObjectNew.(*appsv1alpha1.NIMBuild)
+					if ok {
+						// Handle case where object is marked for deletion
+						if !newNIMBuild.ObjectMeta.DeletionTimestamp.IsZero() {
+							return true
+						}
+
+						// Handle only spec updates
+						return !reflect.DeepEqual(oldNIMBuild.Spec, newNIMBuild.Spec)
+					}
+				}
+				// For other types we watch, reconcile them
+				return true
+			},
+		}).
+		Complete(r)
+}
+
+func (r *NIMBuildReconciler) cleanupNIMBuild(ctx context.Context, nimBuild *appsv1alpha1.NIMBuild) error {
+	var errList []error
+	logger := r.GetLogger()
+
+	// All owned objects are garbage collected
+
+	// Fetch the pod
+	podName := types.NamespacedName{Name: nimBuild.Name + "-engine-build-pod", Namespace: nimBuild.Namespace}
+	pod := &corev1.Pod{}
+	if err := r.Get(ctx, podName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		logger.Error(err, "unable to fetch the pod for cleanup", "pod", podName)
+		return err
+	}
+
+	if err := r.Delete(ctx, pod); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Pod already deleted", "pod", podName)
+			return nil
+		}
+		logger.Error(err, "unable to delete associated pods during cleanup", "job", podName, "pod", pod.Name)
+		errList = append(errList, err)
+	}
+
+	if len(errList) > 0 {
+		return fmt.Errorf("failed to cleanup resources: %v", errList)
+	}
+
+	return nil
+}
+
+func (r *NIMBuildReconciler) reconcileNIMBuild(ctx context.Context, nimBuild *appsv1alpha1.NIMBuild) (reconcile.Result, error) {
+	logger := r.GetLogger()
+	nimCacheNamespacedName := types.NamespacedName{Name: nimBuild.Spec.NIMCacheRef, Namespace: nimBuild.GetNamespace()}
+
+	nimCache := &appsv1alpha1.NIMCache{}
+	if err := r.Get(ctx, nimCacheNamespacedName, nimCache); err != nil {
+		logger.Error(err, "unable to fetch NIMCache for NIMBuild", "NIMCache", nimCacheNamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if nimCache.Status.State == appsv1alpha1.NimCacheStatusReady {
+		err := r.reconcileEngineBuild(ctx, nimBuild, nimCache)
+		if err != nil {
+			logger.Error(err, "reconciliation of nimbuild failed")
+			return ctrl.Result{}, err
+		}
+	} else {
+		conditions.UpdateCondition(&nimBuild.Status.Conditions, appsv1alpha1.NimBuildConditionWaitForNimCache, metav1.ConditionTrue, "NIMCache not ready", "Waiting for NIMCache to be ready before building engines")
+		nimBuild.Status.State = appsv1alpha1.NimBuildStatusPending
+		return ctrl.Result{RequeueAfter: time.Second * 30}, nil
+
+	}
+
+	err := r.updateNIMBuildStatus(ctx, nimBuild)
+	if err != nil {
+		logger.Error(err, "Failed to update NIMBuild status", "NIMBuild", nimBuild.Name)
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *NIMBuildReconciler) reconcileEngineBuild(ctx context.Context, nimBuild *appsv1alpha1.NIMBuild, nimCache *appsv1alpha1.NIMCache) error {
+	logger := r.GetLogger()
+	// If the NIMCache is not in a state to build engines, return early
```

Comment:
nit, these two comments can be removed as the check is already enforced by the caller

Reply:
Fixed